lib/Echidna/RPC.hs

```diff
@@ -0,0 +1,146 @@
+                                              <*> (read <$> (v .: "gas_price"))
+                                              <*> (decode <$> (v .: "data"))
+                                              <*> v .: "value"
+         _ -> M.fail "event should be one of \"AccountCreated\", \"ContractCreated\", or \"FunctionCall\""
+    where decode :: T.Text -> ByteString
+          decode = decodeHex . encodeUtf8 . T.drop 2
+          decodeHex = fst . BS16.decode
+
+
+-- | Handler for parsing errors
+data EthenoException = EthenoException String
+
+instance Show EthenoException where
+    show (EthenoException e) = "Error parsing Etheno initialization file: " ++ e
+
+instance Exception EthenoException
+
+
+-- | Main function: takes a filepath where the initialization sequence lives and returns
+-- | the initialized VM along with a list of Addr's to put in GenConf
+loadEthenoBatch :: (MonadThrow m, MonadIO m, Has TxConf y, MonadReader y m) => [T.Text] -> FilePath -> m (VM, [Addr])
+loadEthenoBatch ts fp = do
+  bs <- liftIO $ eitherDecodeFileStrict fp
+
+  case bs of
+       (Left e) -> throwM $ EthenoException e
+       (Right ethenoInit) -> do
+         -- Separate out account creation txns to use later for config
+         let (accounts, txs) = partition (\case { AccountCreated{} -> True; _ -> False; }) ethenoInit
+             knownAddrs      = map (\(AccountCreated a) -> a) accounts
+
+         -- Execute contract creations and initial transactions,
+         let blank  = vmForEthrunCreation empty & env . contracts .~ fromList []
+             initVM = foldM (execEthenoTxs ts) Nothing txs
+
+         (addr, vm') <- runStateT initVM blank
+         case addr of
+              Nothing -> throwM $ EthenoException "Could not find a contract with echidna tests"
+              Just a  -> do
+                vm <- execStateT (liftSH . loadContract $ a) vm'
+                return (vm, knownAddrs)
+
+-- | Takes a list of Etheno transactions and loads them into the VM, returning the
+-- | address containing echidna tests
+execEthenoTxs :: (MonadState x m, Has VM x, MonadThrow m, Has TxConf y, MonadReader y m) => [T.Text] -> Maybe Addr -> Etheno -> m (Maybe Addr)
+execEthenoTxs ts addr et = do
+  setupEthenoTx et
+  res <- liftSH exec
+  g <- view (hasLens . propGas)
+  case (res, et) of
+       (Reversion,   _)               -> throwM $ EthenoException "Encountered reversion while setting up Etheno transactions"
+       (VMFailure x, _)               -> vmExcept x >> return addr
+       (VMSuccess bc,
+        ContractCreated _ ca _ _ _ _) -> do
+          hasLens . env . contracts . at ca . _Just . contractcode .= InitCode ""
+          liftSH (replaceCodeOfSelf (RuntimeCode bc) >> loadContract ca)
+          og <- get
+          -- See if current contract is the same as echidna test
+          case addr of
+               Just m  -> return $ Just m
+               Nothing -> let txs = ts <&> \t -> Tx (Left (t, [])) ca ca g 0 0 (0,0)
```

Comment:
since we are just calling the echidna test functions i don't think it should really matter who calls the contract, as tests that revert/break on certain contract addresses are bad tests.

Reply:
hmmm, not sure this is true in practice (consider tests of an address-based authorization fn). I think we should just parse the cfg and do it right

Reply:
if we pull in the required `Conf` module then we get some circular dependencies so let's do this afterwards.